File: src/lib/permissions.ts
import { AuthenticatedUser } from './apiAuth';

/**
 * Permission helper utilities for consistent authorization checks.
 * Centralizes permission logic to make changes easier.
 */

/**
 * Check if user is an owner or admin
 */
export function isOwnerOrAdmin(user: AuthenticatedUser): boolean {
  return user.role === 'owner' || user.role === 'admin';
}

/**
 * Check if user can edit a post.
 * Rules: Only the post author or owner/admin can edit.
 */
export function canEditPost(
  user: AuthenticatedUser,
  post: { authorId: string }
): boolean {
  return post.authorId === user.id || isOwnerOrAdmin(user);
}

/**
 * Check if user can delete a post.
 * Rules: Only the post author or owner/admin can delete.
 */
export function canDeletePost(
  user: AuthenticatedUser,
  post: { authorId: string }
): boolean {
  return post.authorId === user.id || isOwnerOrAdmin(user);
}

/**
 * Check if user can delete a comment.
 * Rules: Only the comment author or owner/admin can delete.
 */
export function canDeleteComment(
  user: AuthenticatedUser,
  comment: { authorId: string }
): boolean {
  return comment.authorId === user.id || isOwnerOrAdmin(user);
}

/**
 * Check if user can remove a family member.
 * Rules:
 * - Only owner/admin can remove members
 * - Cannot remove yourself
 * - Cannot remove the owner
 */
export function canRemoveMember(
  actingUser: AuthenticatedUser,
  targetUser: { id: string; role: string }
): {
  allowed: boolean;
  reason?: 'NOT_ADMIN' | 'CANNOT_REMOVE_SELF' | 'CANNOT_REMOVE_OWNER';
} {
  // Must be owner or admin to remove anyone
  if (!isOwnerOrAdmin(actingUser)) {
    return { allowed: false, reason: 'NOT_ADMIN' };
  }

  // Cannot remove yourself
  if (targetUser.id === actingUser.id) {
    return { allowed: false, reason: 'CANNOT_REMOVE_SELF' };
  }

  // Cannot remove the owner
  if (targetUser.role === 'owner') {
    return { allowed: false, reason: 'CANNOT_REMOVE_OWNER' };
  }

  return { allowed: true };
}

File: apps/api/src/permissions.py
from .schemas.auth import UserResponse


def is_owner_or_admin(user: UserResponse) -> bool:
    return user.role in ("owner", "admin")


def can_edit_post(user: UserResponse, post_author_id: str) -> bool:
    return user.id == post_author_id or is_owner_or_admin(user)


def can_delete_comment(user: UserResponse, comment_author_id: str) -> bool:
    return user.id == comment_author_id or is_owner_or_admin(user)


def can_remove_member(current_user: UserResponse, target_user_id: str, target_role: str) -> bool:
    if not is_owner_or_admin(current_user):
        return False
    if target_user_id == current_user.id:
        return False
    if target_role == "owner":
        return False
    return True
